src/java-opentimelineio/src/test/java/io/opentimeline/GeneratorReferenceTest.java

```diff
@@ -0,0 +1,104 @@
+                        new TimeRange(
+                                new RationalTime(0, 24),
+                                new RationalTime(100, 24)))
+                .setParameters(parameters)
+                .setMetadata(metadata)
+                .build();
+    }
+
+    @Test
+    public void testConstructor() {
+        assertEquals(generatorReference.getGeneratorKind(), "SMPTEBars");
+        assertEquals(generatorReference.getName(), "SMPTEBars");
+        assertEquals(generatorReference.getParameters().get("test_param").safelyCastDouble(), 5.0);
+        assertEquals(generatorReference.getParameters().size(), 1);
+        assertEquals(generatorReference.getMetadata().get("foo").safelyCastString(), "bar");
+        assertEquals(generatorReference.getMetadata().size(), 1);
+        assertTrue(generatorReference.getAvailableRange()
+                .equals(
+                        new TimeRange(
+                                new RationalTime(0, 24),
+                                new RationalTime(100, 24))));
+    }
+
+    @Test
+    public void testSerialize() {
+        try {
+            Any refAny = new Any(generatorReference);
+            ErrorStatus errorStatus = new ErrorStatus();
+            String encoded = Serialization.serializeJSONToString(refAny, errorStatus);
+            SerializableObject decoded = SerializableObject.fromJSONString(encoded, errorStatus);
+            assertTrue(decoded.isEquivalentTo(generatorReference));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    @Test
+    public void testReadFile() {
+        String projectRootDir = System.getProperty("user.dir");
+        String sampleDataDir = projectRootDir + File.separator +
+                "src" + File.separator + "test" + File.separator + "sample_data";
+        String genRefTest = sampleDataDir + File.separator + "generator_reference_test.otio";
+        File file = new File(genRefTest);
+        assertTrue(file.exists());
+        ErrorStatus errorStatus = new ErrorStatus();
+        Any destination = new Any(new SerializableObject());
+//        assertTrue(Deserialization.deserializeJSONFromFile(genRefTest, destination, errorStatus));
+        SerializableObject serializableObject = SerializableObject.fromJSONFile(genRefTest, errorStatus);
+        Timeline timeline = new Timeline(serializableObject);
+        Stack stack = timeline.getTracks();
+        List<SerializableObject.Retainer<Composable>> tracks = stack.getChildren();
+//        Composable composable = tracks.get(0).value();
+//        System.out.println(composable.className);
+//        Track track = new Track(tracks.get(0).value());
+//        ArrayList<SerializableObject.Retainer<Composable>> track0Children = new ArrayList<>(track.getChildren());
+//        Clip clip = new Clip(track0Children.get(0).value());
+//        assertEquals(((GeneratorReference) clip.getMediaReference()).getGeneratorKind(), "SMPTEBars");
+    }
```

Comment:
@margant @reinecke what do you think would be the best way to handle generics here.

I'll elaborate the problem a little:

This is a part of `SerializableObject.java`:

```java
public class SerializableObject extends OTIONative {
    ...
    public static class Retainer<T extends SerializableObject> extends OTIONative {

        public Retainer(long nativeHandle) {
            this.nativeHandle = nativeHandle;
        }

        public Retainer(T value) {
            this.initialize(value);
        }

        public Retainer(Retainer<T> retainer) {
            this.initialize(retainer.value());
        }

        private native void initialize(T value);

        public native T value();

        public native T takeValue();

        private native void dispose();

        @Override
        public void close() throws Exception {
            dispose();
        }
    }

    ...
}
```

Here Retainer<> is a generic with `T extends SerializableObject`.

So all the objects in the `List<Retainer>` on line 84 will actually be `Retainer<SerializableObject>`. But when I try to get the value from the Retainer in line 85, it crashes because it won't be able to downcast the `SerializableObject` to a `Composable`. 

One solution could be to change the `value()` function definition in the Retainer class to return `new T(value)` instead of just the native method used right now, and add constructors in each derived class of SerializableObject that take a SerializableObject as a parameter and set the nativeHandle.

Is there a better way to handle this?

PS: It seems like it's going to take a lot more time to get these tests done than expected. I'm getting unexpected `bad_alloc`s here too. :/

Reply:
@margant any thoughts on this?

Reply:
Looking now, I am still trying to see how C++ returns SerializableObject in the first place. 
I see that the Composable got type-erased because your track is of type Serializable$Retainer


```
JNIEXPORT jobjectArray JNICALL
Java_io_opentimeline_opentimelineio_Composition_getChildrenNative(
    JNIEnv* env, jobject thisObj)
{
    auto thisHandle = getHandle<OTIO_NS::Composition>(env, thisObj);
    std::vector<OTIO_NS::SerializableObject::Retainer<OTIO_NS::Composable>>
        result = thisHandle->children();
    return composableRetainerVectorToArray(
        env,
        *(new std::vector<
            OTIO_NS::SerializableObject::Retainer<OTIO_NS::Composable>>(
            result)));
}
```

Reply:
Out of curiosity, do we need to expose the `Retainer` generic up to the Java code? I think that construct has more to do with C++ memory management than anything, is it possible the `Retainer` can be limited to under-the-hood use in the JNI code?

In other words, if the Java objects are keeping pointers to the `Retainer` under the hood, then when the java object is GC'd, the retainer can be let go at that point (in the phantom object) and the retainer can decrement the ref counting.

Please let me know if I'm missing something, I haven't done the deep research into whether or not there are specific issues with what I'm describing.

Reply:
@reinecke was thinking along the same lines, rather then having Retainer<Composable> if we could have Composable have Retainer object privately inside. I think this could solve type erasure problem as well. @KarthikRIyer wdyt?

Reply:
Yes, this should solve the type erasure problem. This makes sense. I'll try this and let you know how it goes.

Reply:
Till now I was trying for a literal translation of C++. But this seems much better.

Reply:
@KarthikRIyer For this project we like to have language bindings feel as "correct" as possible for the language they are exposed in. This sometimes means that the bindings layer will be more work to maintain, but we are willing to accept that to make the library as friendly to the users as it can be.
So if you encounter anything in the API that doesn't feel right in Java, please call it out and we can work through trying to address it.

It feels like things are coming along nicely!

Reply:
I agree. Even if it is more work, usability should be the primary focus. I've tried to keep things Java like as much as possible, apart from this Generics thing. If there are more avenues for improvement I'll discover them while writing tests. 
I'll ask if I anything doesn't feel right. Thanks!

Reply:
Awesome, thanks @KarthikRIyer. Absolutely agree with @reinecke , for users of Java environment are likely used to an already established paradigm and the more native we can make this feel, the easier it will be to use. 

I added comments throughout the Java codebase with some questions/comments that might make it more "native". Hopefully they make sense, but happy to discuss if anything needs clarification.